### PR TITLE
Improvements in track quality criteria in PF for muon seeded iterations (76X)

### PR DESCRIPTION
**Pull request by bachtis**
See #11433

### CODE REVIEW

File: RecoParticleFlow/PFProducer/plugins/importers/ImprovedTracksImporter.cc [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+    if (_debug) std::cout << " PFBlockAlgo: skip badly measured track"
+		     << ", P = " << P 
+		     << ", Pt = " << Pt 
+		     << " DPt = " << DPt 
+		     << ", N(hits) = " << NHit << " (Lost : " << LostHits << "/" << NLostHit << ")"
+		     << ", Algo = " << Algo
+		     << std::endl;
+    if (_debug) std::cout << " cut is DPt/Pt < " << _DPtovPtCut[Algo] * sigmaHad << std::endl;
+    if (_debug) std::cout << " cut is NHit >= " << _NHitCut[Algo] << std::endl;
+    /*
```
Comment (slava77) [inline RecoParticleFlow/PFProducer/plugins/importers/ImprovedTracksImporter.cc]: please cleanup commented out parts ... if needed keep behind LogTrace or LogDebug (preferred)

File: RecoParticleFlow/PFProducer/plugins/importers/ImprovedTracksImporter.cc [hunk 2]
```diff
@@ -0,0 +1,259 @@
+#include "RecoParticleFlow/PFProducer/interface/BlockElementImporterBase.h"
+#include "DataFormats/ParticleFlowReco/interface/PFBlockElementTrack.h"
+#include "DataFormats/ParticleFlowReco/interface/PFRecTrackFwd.h"
+#include "DataFormats/ParticleFlowReco/interface/PFRecTrack.h"
+#include "DataFormats/TrackReco/interface/Track.h"
+#include "DataFormats/MuonReco/interface/MuonFwd.h"
+#include "DataFormats/MuonReco/interface/Muon.h"
+#include "RecoParticleFlow/PFProducer/interface/PFMuonAlgo.h"
+
+class ImprovedTracksImporter : public BlockElementImporterBase {
```
Comment (slava77) [inline RecoParticleFlow/PFProducer/plugins/importers/ImprovedTracksImporter.cc]: "Improved" is generally not a good name
What will happen to the next developments? "ImprovedImproved"?

**Comment by lgray** [comment]
@bachtis Not possible to extend the existing general tracks importer? (to avoid these codes from diverging over time?)

File: RecoParticleFlow/PFProducer/plugins/importers/ImprovedTracksImporter.cc (hunk 2 above)
Comment (lgray) [inline RecoParticleFlow/PFProducer/plugins/importers/ImprovedTracksImporter.cc]: Clearly, "BetterImproved", followed by "UpgradedBetterImproved"...

**Comment by bachtis** [comment]
@lgray  No because then HLT cfg needs reparsing Read the description before you make comments
@lgray , @slava77  I like the name and I couldnt think of a better one

**Comment by lgray** [comment]
@bachtis I read your description, that says it is inconvenient rather than impossible, which is why I made the comment. 

It is not good to have two sets of code that do mostly the same thing. Since it is going to 75X there is likely time to reparse the HLT configs.

However, if time *is* pressing. This can be taken care of in another PR.

@slava77 @Martin-Grunewald comments?

**Comment by lgray** [comment]
Also, what about GeneralTracksImporterV2? Much less wordy and has the benefit of not needing a countably infinite list of adjectives.

**Comment by bachtis** [comment]
Well if @Martin-Grunewald agrees with reparsing we can replace the old one. The matrix tests will crash at HLT if this is not done though .

**Comment by lgray** [comment]
@bachtis Make a customization function named customizeFor_<PRNumber> and stick it in HLTrigger/Configuration. There are a few examples in there. This will allow matrix tests to run.

**Comment by bachtis** [comment]
OK then lets see if it is OK to change the hltParticleFlowBlock module.and then decide how to proceed

**Comment by davidlange6** [comment]
the hlt is supporting changes to 75x via customize functions. Should be easy

**Comment by bachtis** [comment]
so can somebody tell me what exactly I have to do- which files to change? In that case I would prefer to modify the old importer instead of adding new one

**Comment by Martin-Grunewald** [comment]
If you'd use a fillDescriptions method, you can add all the parameters you want,
without HLT parsing, if the default values are fine for HLT. No need for customisation.

**Comment by Martin-Grunewald** [comment]
@davidlange6 
Why are we allowing new code in without fillDescriptions?
(Cf ORP wiki blurb?)

**Comment by davidlange6** [comment]
An even better solution.



On Sep 23, 2015, at 5:26 PM, Martin Grunewald <notifications@github.com<mailto:notifications@github.com>> wrote:


@davidlange6<https://github.com/davidlange6>
Why are we allowing new code in without fillDescriptions?
(Cf ORP wiki blurb?)

**Comment by lgray** [comment]
The python change for this happens in some PSet inside of a VPSet. Is there a way to do fillDescriptions for an individual plugin in a list?

**Comment by bachtis** [comment]
HI . I am running the matrix. In fact since HLT iterations are different it might work out of the box .

**Comment by bachtis** [comment]
In fact no intervention is needed in HLT config . Implemented code review comments. PR should be updated now

**Comment by Dr15Jones** [comment]
Yes one can have a fillDescription state the properties of PSets within a VPSet.

File: RecoParticleFlow/PFProducer/plugins/importers/GeneralTracksImporter.cc [hunk 3]
```diff
@@ -33,6 +38,9 @@ class GeneralTracksImporter : public BlockElementImporterBase {
   const std::vector<double> _DPtovPtCut;
   const std::vector<unsigned> _NHitCut;
   const bool _useIterTracking,_cleanBadConvBrems,_debug;
+
+  PFMuonAlgo *pfmu_;
```
Comment (slava77) [inline RecoParticleFlow/PFProducer/plugins/importers/GeneralTracksImporter.cc]: unique_ptr indeed, please.

**Comment by bachtis** [comment]
unique_ptr done

**Comment by lgray** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@cmsbuild please test
[I suppose migration to fillDescriptions is not happening, since now it's an "old code" fix]

**Comment by lgray** [comment]
@slava77 already asked for test..

**Comment by slava77** [comment]
I'm running higher stat tests to see the effects.
Code looks ok (no changes needed from visual inspection)

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/11434	bc37c82
- track cleaning updates: expect less charged hadrons
- jenkins tests pass and comparisons with baseline show no differences (the phase space of changes is not sampled in the tests; it is small)
- tested locally in CMSSW_7_6_X_2015-09-22-1500 : changes are observed in higher stat tests with high-pt tracks occurring naturally: in wf 1313 (3 TeV dijet) and muon 1 TeV gun 
    - changes in wf1313 are more significant and are in a more natural environment:
        - 22/400 (~5%)  events have differences
        -  about a half of events have  larger MET ( 4 larger by 20%; 3 events have MET going smaller by 20%). So, relative to MET in this sample the change isn't making an improvement.
        - of all PFCandidates in the event: there are fewer high-pt charged hadrons (-22) and muons  (-1) and more photons (+14) and neutral hadrons (+13). This migration is expected if a large energy is present.
    - muon 1 TeV sample doesn't have particularly large changes
- ==> things look OK in the "normal" events. For the signoff I'm relying more  on observations from the more rare events cases used to make this PR.

**Comment by davidlange6** [comment]
+1

**Comment by lgray** [comment]
@bachtis Does the migration observed in wf1313 have any sizeable impact on the resolution of high pT jets?

**Comment by lgray** [comment]
@bachtis Just asking since there are no plots on the PR...

**Comment by bachtis** [comment]
Should not have  any impact . If you have a ~TeV track that will be killed the calorimeter should measure it better . Only one event changed in the Flat QCD sample I rerecoed so I dont expect any effect

**Comment by slava77** [comment]
I could rerun on more events. There wasn't much to see with 22 events
changing only.

**Comment by lgray** [comment]
@bachtis Sure, that I can believe, but need the plot. My being curious is simply healthy skepticism.

@slava77 If we could put a plot here it would help. (I guess 1k events or something?)

**Comment by slava77** [comment]
@lgray I don't think 1K events will be enough (4K maybe).
Here are a couple of plots from the tests done already (from wf 1313)

First is to show that tracks go away on the high end of chFrac (this is a more clear "thing are in the right direction" plot)
wf1313_ak4pf_chfrac

And the reco/gen plot for high et jets: it's not very convincingly mildly better
wf1313_ak4pf_recoogen

**Comment by lgray** [comment]
@slava77 @bachtis Ahh, OK this is certainly in the noise concerning the resolution. Thanks.

**Pull request merged**